poc/test_sigma_protocols.sage

```diff
@@ -305,6 +305,181 @@ def test_and_composition():
+            # subtracting the prover's shares from the verifier challenge.
+            for challenge_share, sim_responses in unknown_witness_prover_states:
+                known_state_challenge -= challenge_share
+                responses.append(sim_responses)
+                challenges.append(challenge_share)
+
+            # Include the response for the known protocol at the correct index
+            # (i.e., the index of the protocol in the original list of protocols)
+            (known_prover_state, known_index) = known_prover_states[0]
+            known_response = self.protocols[known_index].prover_response(known_prover_state, known_state_challenge)
+
+            responses.insert(known_index, known_response)
+            challenges.insert(known_index, known_state_challenge)
+
+            return (challenges[:-1], responses)
+
+        def verifier(self, commitments, challenge, _response):
+            challenges, responses = _response
+            assert len(commitments) == len(responses)
+            last_challenge = challenge - sum(challenges)
+            challenges.append(last_challenge)
+            assert all(
+                protocol.verifier(commitment, challenge, response)
+                for protocol, commitment, challenge, response in zip(self.protocols, commitments, challenges, responses)
+            )
+
+            return True
+
+        def serialize_batchable(self, commitments, challenge, _response):
+            challenges, responses = _response
+            return b''.join(
+                    [protocol.serialize_batchable(commitment, challenge, response) for protocol, commitment, response in zip(self.protocols, commitments, responses)] +
+                    [protocol.instance.Domain.serialize([challenge]) for (protocol, challenge) in zip(self.protocols[:-1], challenges)]
+                )
+
+        def deserialize_batchable(self, proof_string):
+            commitments = []
+            challenges = []
+            responses = []
+
+            start = 0
+            for protocol in self.protocols:
+                proof_len = protocol.instance.commit_bytes_len + protocol.instance.response_bytes_len
+                commitment, response = protocol.deserialize_batchable(proof_string[start : start + proof_len])
+                start += proof_len
+                commitments.append(commitment)
+                responses.append(response)
+
+            # The last part of the proof string is the challenges
+            for protocol in self.protocols[:-1]:
+                # we should not be needing to inspect Domain here
+                challenge_len = protocol.instance.Domain.scalar_byte_length()
+                challenge = protocol.instance.Domain.deserialize(proof_string[start : start + challenge_len])
+                challenges.append(challenge[0])
+                start += challenge_len
+
+            return (commitments, challenges, responses)
```

Comment:
I would expect `deserialize_batchable` to be the inverse of `serialize_batchable`, and that (by correctness) its output verifies. Can you pack again `challenges` with `responses`?